content/blog/2019-09-25-manually-managed-memory.md

```diff
@@ -0,0 +1,303 @@
+
+In an interpreter setting, where we can rely on another language runtime to do
+the real physical memory allocation and garbage collection for us, this is not
+a terribly complex addition to Bril. However, there were a few small details
+that were tricky to get exactly right.
+
+Firstly, we had to handle how to parse and represent new types that are
+parameterized on other types. We wanted to syntactically enforce, with the
+parser, that pointer types had to be fully specified.
+
+The original type parsing specification looked like this:
+
+```C
+type: CNAME
+```
+
+And created an AST node where the type was specified as a `string`.
+
+Our new version looks like this:
+```C
+type: "ptr<" ptrtype ">" | basetype
+ptrtype: type
+basetype: CNAME
+```
+
+And we create an AST node where the type is specified as a (potentially nested)
+object with one field named "ptr". For example, the AST representation of a node
+with type `ptr<bool>` looks like
+
+```C
+type: { ptr: "bool" }
+```
+
+We could have decided to maintian the pointer type abstract representation as
+a string, but that would require re-parsing that type string repeatedly
+throughout the interpreter. While we avoided that problem, we now had to deal
+with the annoyance that some types were represented as strings and others were
+JSON objects. This lead to a fair bit of refactoring and slightly tricky runtime
+typechecking, but we decided it was worth it compared to having to do
+sophisticated string pattern matching at runtime.
+
+### Evaluation
+
+We evaluated our implementation through qualitative testing rather than
+quantitative measurement. While one might argue we should measure memory
+allocation performance, we would really just be measuring how well the
+javascript runtime allocates memory and garbage collects.
+
+We wanted to evaluate the correctness of our code and
+see if we threw reasonable errors under all erroneous conditions. We created
+a number of test cases that stress pointer arithmetic (similar to the large
+example presented earlier in this post). Key features to test were:
+
+ - Allocating memory of various sizes
+ - Reading & writing memory
+ - Re-writing memory
+ - Ensuring that pointers to pointers function correctly
+
+Additionally, we needed to check a number of "bad" cases, which we expected the
+interpreter to catch and report as errors with reasonable error messages:
+
+ - Passing non-pointers to `load`, `store` or `free` operations
+ - Allocating pointers with non-positive size
+ - Failing to free memory by the end of the program
+ - Freeing the same allocation multiple times
+ - Trying to free a pointer into the middle of an allocation region
+ - Accessing memory "out of bounds" of a given access
+ - Writing the wrong type of data into a pointer (e.g. store `int` into
+   a `ptr<bool>`)
+   - N.B. that "reading" data of the wrong type is still allowed, which actually
+     mirrors the current interpreter implementation for other operations
```

Comment:
Ah yeah, good point—the interpreter doesn't actually care about the types. Maybe worth mentioning that explicitly: that our reference interpreter implementation doesn't try to do (much) dynamic type checking—it can do arbitrarily bad things on ill-typed programs.